### ISSUE TITLE
Train to upscale low quality JPEG 

### ISSUE THREAD

**Issue by 3zhang** [nagadomi/nunif#75]
My goal is to train/fine-tune the waifu2x model to upscale low quality jpeg photos (quality factor around 70). How do I set the training parameter? Is `--da-jpeg-p` related?

**Comment by nagadomi** [comment]
waifu2x already has GAN based denoise+upscale models for photo.
You can check the results of the pre-training model for each noise level at https://unlimited.waifu2x.net/
(select `swin_unet/photo` for `Model` option.)

---
waifu2x (art) models has pretrained models for each noise level.
```
noise level 0(Low):  JPEG quality 85~95
noise level 1(Middle):  JPEG quality 65~85, +noise level 0
noise level 2(High):  JPEG quality 27~70, +noise level 0, 1
noise level 3(Highest):  JPEG quality 27~70
```
They are separated by `--noise-level  [0-3]` option.
For example, use `--method noise_scale --noise-level 1` option for 2x noise-level 1.
JPEG quality 70 corresponds to noise-level 1, but in my experience, noise-level 3 is the best look.
You can check the results for each noise-level at the above website.

For training photo models, use `--style photo` option (`--style art` by default).
When `--style photo` is specified, additional noise types related to photo, such as grain noise, will be added in addition to JPEG noise.

The options I used for pretrained models can be found at https://github.com/nagadomi/nunif/tree/master/waifu2x/appendix

**Comment by 3zhang** [comment]
> waifu2x already has GAN based denoise+upscale models for photo. You can check the results of the pre-training model for each noise level at https://unlimited.waifu2x.net/ (select `swin_unet/photo` for `Model` option.)
> 
> waifu2x (art) models has pretrained models for each noise-level.
> 
> ```
> noise-level 0(Low):  JPEG quality 85-95
> noise-level 1(Middle):  JPEG quality 65-85, +noise level 0
> noise-level 2(High):  JPEG quality 27-70, +noise level 0, 1
> noise-level 3(Highest):  JPEG quality 27-70
> ```
> 
> They are separated by `--noise-level [0-3]` option. For example, use `--method noise_scale --noise-level 1` option for 2x noise-level 1. JPEG quality 70 corresponds to noise-level 1, but in my experience, noise-level 3 is the best look. You can check the results for each noise-level at the above website.
> 
> For training photo models, use `--style photo` option (`--style art` by default). When `--style photo` is specified, additional noise types related to photo, such as grain noise, will be added in addition to JPEG noise.
> 
> The options I used for pretrained models can be found at https://github.com/nagadomi/nunif/tree/master/waifu2x/appendix

Thanks. So which arch should I use in training?

**Comment by nagadomi** [comment]
Current answers is: `waifu2x.swin_unet_2x` for 2x, `waifu2x.swin_unet_4x` for 4x.
You can find these pre-trained models under `waifu2x/pretrained_models/swin_unet/`.

I am currently developing a new arch in the `dev` branch and it can achieve PSNR +1 for art 2x.
However, it is still under development and the pre-training model will be released 1-2 months later.

**Comment by 3zhang** [comment]
> Current answers is: `waifu2x.swin_unet_2x` for 2x, `waifu2x.swin_unet_4x` for 4x. You can find these pre-trained models under `waifu2x/pretrained_models/swin_unet/`.
> 
> I am currently developing a new arch in the `dev` branch and it can achieve PSNR +1 for art 2x. However, it is still under development and the pre-training model will be released 1-2 months later.

I want to 2x upscale. Can I train `waifu2x.swin_unet_2x` with  noise-level 1? Which loss should I use?

**Comment by nagadomi** [comment]
>Can I train waifu2x.swin_unet_2x with noise-level 1?

Yes.
For noise-level, I think it is better to use the option value that you feel is the best result from the above website.

>Which loss should I use?

The pre-training model was trained with `--loss lbp5`(`--loss lbp` for 2x) for PSNR optimized model and `--loss l1lpips --discriminator l3v1c` for GAN model (see `train_photo_psnr.sh` and `train_photo_gan.sh` ).
However, the recently added `--loss y_l1fftgrad` should be faster and gives better results than `--loss lbp` and `--loss lbp5` for PSNR optimized model.

**Comment by 3zhang** [comment]
> > Can I train waifu2x.swin_unet_2x with noise-level 1?
> 
> Yes. For noise-level, I think it is better to use the option value that you feel is the best result from the above website.
> 
> > Which loss should I use?
> 
> The pre-training model was trained with `--loss lbp5`(`--loss lbp` for 2x) for PSNR optimized model and `--loss l1lpips --discriminator l3v1c` for GAN model (see `train_photo_psnr.sh` and `train_photo_gan.sh` ). However, the recently added `--loss y_l1fftgrad` should be faster and gives better results than `--loss lbp` and `--loss lbp5` for PSNR optimized model.

I'm a bit confused about the `size` parameter. What's the difference between the `size` in create_training_data.py and `size` in train.py?

**Comment by nagadomi** [comment]
`--size` option in train.py is the image(patch/tile) size to input for the model. For swin_unet arch, `--size 64` is recommended. (112 is also possible, but very slow.)

`--size` option in create_training_data.py is the image size for the dataset.
During training, input images are generated with random cropping from the dataset images.
For 4x training it must be higher than `--size 320`.
```
scale_factor = 4
required_margin = 16 * scale_factor
size_in_train = 64
upscale_target_size = size_in_train * scale_factor # 256
required_dataset_image_size = upscale_target_size + required_margin # 320
```
For 2x training it must be higher than `--size 160`.
If you are using fast SSD disk, 640 is recommended.
However, 320 is also fine (it speeds up the image loading).

**Comment by 3zhang** [comment]
> `--size` option in train.py is the image(patch/tile) size to input for the model. For swin_unet arch, `--size 64` is recommended. (112 is also possible, but very slow.)
> 
> `--size` option in create_training_data.py is the image size for the dataset. During training, input images are generated with random cropping from the dataset images. For 4x training it must be higher than `--size 320`.
> 
> ```
> scale_factor = 4
> required_margin = 16 * scale_factor
> size_in_train = 64
> upscale_target_size = size_in_train * scale_factor # 256
> required_dataset_image_size = upscale_target_size + required_margin # 320
> ```
> 
> For 2x training it must be higher than `--size 160`. If you are using fast SSD disk, 640 is recommended. However, 320 is also fine (it speeds up the image loading).

Just to be sure. If `--num-samples` is smaller than the number of images in the train folder, will the program only feed the first `num-samples` images to the model? Is the image order shuffled on each epoch?

**Comment by nagadomi** [comment]
Train data loader uses random sampler. Each input image is randomly selected from all training images.
For eval images, --num-samples option is not used and all images are used in one epoch.

**Comment by nagadomi** [comment]
It is difficult to give advice for GAN model training.
I myself don't feel that is working very well.

Here is a small working example. `--data-dir` should be changed for your env.
```
DEBUG=1 python train.py waifu2x --arch waifu2x.swin_unet_2x --data-dir ./data/photo --model-dir ./models/gan_2x_test --method noise_scale --noise-level 3 --style photo --loss l1lpips --discriminator l3v1c --size 64 --batch-size 16 --hard-example none --update-criterion all --learning-rate 0.00003 --scheduler step --max-epoch 4 --checkpoint-file ./waifu2x/pretrained_models/swin_unet/art/noise3_scale2x.pth
```
If you are using Windows, you need to set environment variables with `set DEBUG=1` command.
Also, there was a bug with a recent change I made, I fixed it https://github.com/nagadomi/nunif/commit/9b21a91ebe232807dbf1a476d309d15bae01cb4d. If you get an error, update the code.

The above command takes only a few minutes because `--max-epoch 4` is specified.
You can check the upscaling results of the trained model with the following command.
```
python -m waifu2x.cli -m noise_scale --noise-level 3 --model-dir ./models/gan_2x_test -i tmp/test_images -o ./tmp/gan_2x_out/ --yes
```
`-i tmp/test_images` is the directory containing the test images.


Basics
- GAN training must start with a pretrained model. In the above example, art psnr model is used, but it is better to use photo psnr model. (but i have not released it)
- It is normal that PSNR goes down. However, if it is too low, some artifact may be occurring.

When `DEBUG=1` environment variable is set, the following line will be printed.
```
[   DEBUG] recon: 0.9735, gen: 0.2725, disc: 0.8867, weight: 0.765913, disc skip: 0.033
[   DEBUG] recon: 0.8631, gen: 0.2788, disc: 0.8994, weight: 0.755511, disc skip: 0.001
[   DEBUG] recon: 0.8291, gen: 0.1819, disc: 0.9214, weight: 0.848656, disc skip: 0.0
[   DEBUG] recon: 0.8302, gen: 0.4509, disc: 0.8535, weight: 0.652829, disc skip: 0.116
...
```
```
recon: reconstruction loss
gen: generator loss
disc: discriminator loss
weight: discriminator weight
```

When training is stable,
- `disc` wave around 0.5-0.9

Failing cases
- `disc` keeps decreasing. If `disc` is less than 0.2, GAN training is failing.
- `weight` keeps decreasing. If `weight` is less than 1e-5, the trained model will be high PSNR model

**Comment by nagadomi** [comment]
That is also a difficult question to answer.
Typically, perceptual quality metric such as FID and LIPIS are used for GAN model evaluation.

However, we currently use L1+LPIPS loss, so if minimizing LPIPS or PSNR is the goal, Discriminator is not needed.
We expect GAN training will give better results than directly minimizing L1LPIPS.
So personally, I think it should not be used as stop criteria.

I did not define stop criteria, I defined a training schedule.
- Enough iteration number to train the Discriminator
- Learning rate schedule

If you want to use some criteria to model selection, I would recommend a metric other than LPIPS.

Also you may be concerned about overfitting, but waifu2x models are small and it is basically underfitting. So I am not concerned about overfitting for waifu2x models.
